snuba/cli/subscriptions.py

```diff
@@ -0,0 +1,195 @@
+from datetime import timedelta
+from typing import Optional, Sequence
+
+import click
+
+from snuba import settings
+from snuba.datasets.factory import DATASET_NAMES, enforce_table_writer, get_dataset
+from snuba.environment import setup_logging, setup_sentry
+from snuba.redis import redis_client
+from snuba.subscriptions.consumer import TickConsumer
+from snuba.subscriptions.data import PartitionId
+from snuba.subscriptions.executor import SubscriptionExecutor
+from snuba.subscriptions.scheduler import SubscriptionScheduler
+from snuba.subscriptions.store import RedisSubscriptionDataStore
+from snuba.subscriptions.worker import (
+    SubscriptionResultCodec,
+    SubscriptionWorker,
+)
+from snuba.util import create_metrics
+from snuba.utils.codecs import PassthroughCodec
+from snuba.utils.streams.batching import BatchingConsumer
+from snuba.utils.streams.kafka import (
+    CommitCodec,
+    KafkaConsumer,
+    KafkaProducer,
+    build_kafka_consumer_configuration,
+)
+from snuba.utils.streams.synchronized import SynchronizedConsumer
+from snuba.utils.streams.types import Topic
+
+
+@click.command()
+@click.option(
+    "--dataset",
+    "dataset_name",
+    default="events",
+    type=click.Choice(DATASET_NAMES),
+    help="The dataset to target",
+)
+@click.option("--topic")
+@click.option("--partitions", type=int)
+@click.option("--commit-log-topic")
+@click.option(
+    "--commit-log-group",
+    "commit_log_groups",
+    multiple=True,
+    default=["snuba-consumers"],
+)
+@click.option(
+    "--consumer-group",
+    default="snuba-subscriptions",
+    help="Consumer group use for consuming the dataset source topic.",
+)
+@click.option(
+    "--auto-offset-reset",
+    default="error",
+    type=click.Choice(["error", "earliest", "latest"]),
+    help="Kafka consumer auto offset reset.",
+)
+@click.option(
+    "--bootstrap-server",
+    "bootstrap_servers",
+    multiple=True,
+    help="Kafka bootstrap server(s) to use.",
+)
+@click.option(
+    "--max-batch-size",
+    default=settings.DEFAULT_MAX_BATCH_SIZE,
+    type=int,
+    help="Max number of messages to batch in memory before writing to Kafka.",
+)
+@click.option(
+    "--max-batch-time-ms",
+    default=settings.DEFAULT_MAX_BATCH_TIME_MS,
+    type=int,
+    help="Max length of time to buffer messages in memory before writing to Kafka.",
+)
+@click.option("--schedule-ttl", type=int, default=60 * 5)
+@click.option("--result-topic")
+@click.option("--log-level", default=settings.LOG_LEVEL, help="Logging level to use.")
+def subscriptions(
+    *,
+    dataset_name: str,
+    topic: Optional[str],
+    partitions: Optional[int],
+    commit_log_topic: Optional[str],
+    commit_log_groups: Sequence[str],
+    consumer_group: str,
+    auto_offset_reset: str,
+    bootstrap_servers: Sequence[str],
+    max_batch_size: int,
+    max_batch_time_ms: int,
+    schedule_ttl: int,
+    result_topic: Optional[str],
+    log_level: str,
+) -> None:
+    assert result_topic is not None
```

Comment:
I don't think there is a uniformly reasonable default to put here, unless we want subscriptions for all datasets to be routed to the same topic. I guess a reasonable alternative would be to put this on the `Dataset`. I'd like to eventually encapsulate the contents of `snuba.subscriptions` better and expose that as a `dataset.subscriptions` (e.g. `dataset.subscriptions.create(subscription)`, etc.) but that's not strictly required.

Reply:
Actually, I guess this should go on the dataset so that this can be part of the bootstrap process.

Reply:
I think this (making this optional by providing the subscription result topic via the dataset) will need to be done later, doing it correctly means moving the `KafkaTopicSpec` object out of the dataset storage module and into streams, and I'd rather take the time to do that correctly instead of rush that out (since this isn't super straightforward to use in development yet anyway): https://github.com/getsentry/snuba/blob/3378bb1ec69525669669750be0d0d3311a3ae128/snuba/datasets/table_storage.py#L16-L20

On the plus side, this probably means taking another look at #746, #747, and #751, all of which felt like hacks anyway.